### ISSUE TITLE
Runtimes disagree on semantics of "async" functions

### ISSUE THREAD

**Issue by matanlurey** [dart-lang/sdk#33330]
Customer issue: https://github.com/dart-lang/angular/issues/1375

_Dart VM version: 2.0.0-dev.59.0_

This (nearly) caused a P0 issue on an internal customer (they caught it early). Basically, the semantics of an `async` function in Dart2JS and DDC (and the VM) are not consistent, to be point where DDC hid an infinite loop failure that would occur in Dart2JS (and the VM).

Consider the following:

```dart
import 'dart:async';

import 'package:meta/dart2js.dart' as dart2js;

void main() {
  runZoned(() {
    new AsyncDoEvent().doEvent();
  }, zoneSpecification: new ZoneSpecification(
    scheduleMicrotask: (self, parent, zone, fn) {
      print('scheduleMicrotask()');
      return parent.scheduleMicrotask(zone, fn);
    },
  ));
}

abstract class HasEvent {
  void doEvent();
}

class AsyncDoEvent implements HasEvent {
  @dart2js.noInline
  void doEvent() async {
    print('doEvent()');
  }
}
```

(The `dart2js.noInline` is an attempt to avoid the function being inlined strangely)

### DDC

```
doEvent()
```

### Dart2JS

```
doEvent()
scheduleMicrotask()
```

### VM with `--preview-dart-2`

```
doEvent()
scheduleMicrotask()
```

-----

Notice that only _DDC_ does not schedule a micro-task.

The customer issue is related to how AngularDart schedules change detection and the semantics of a lifecycle event called `ngDoCheck`, where the method is called over and over (indefinitely), starving the microtask queue, and effectively crashing the tab/browser. It does not happen in DDC, however.

-----

A few questions:

* [ ] What does the specification say?
* [ ] If DDC is wrong, is it intentional (perf?) or should it be fixed?

**Comment by matanlurey** [comment]
Sample app is also located here: https://github.com/matanlurey/dart.scratch/tree/master/web_compilers_async

**Comment by vsmenon** [comment]
@lrhn Can you please confirm desired behavior?

@jmesserly - fyi

**Comment by jmesserly** [comment]
Could this be related to sync async flag? @matanlurey do you know if it was enabled or disabled in your examples?

I see this code in DDC:
```
  if (dart.startAsyncSynchronously) {
    body();
  } else {
    scheduleMicrotask(body);
  }
```

I didn't work on the "sync async" feature so I'm not sure what its status is.

**Comment by matanlurey** [comment]
> Could this be related to sync async flag? 

I thought that only effects whether the `body` is entered asynchronously, not if there is eventually an inner-body micro-task. For example, without sync-async, I'd expect to see:

> scheduleMicrotask()
> doEvent()
> scheduleMicrotask()

I never configured anything, so I'm running on the default (in google3 and externally).

**Comment by jmesserly** [comment]
It looks like the VM is executing `doEvent()` synchronously but scheduling completion of the future asynchronously. I don't know what the intended behavior is. I'm not sure who implemented the DDC impl, perhaps @vsmenon remembers? It looks pretty straightforward to switch DDC to match VM's behavior.

It's interesting that all (external/internal) tests are passing with DDC's current implementation, because if I'm understanding this correctly, it's a pretty big difference in timing behavior.

**Comment by matanlurey** [comment]
@jmesserly:

> It's interesting that all (external/internal) tests are passing with DDC's current implementation, because if I'm understanding this correctly, it's a pretty big difference in timing behavior.

I think its a pretty minor spec difference that isn't common:
* An `async` function with effectively no asynchronous body (no `await`)
* Running in a zone that has side-effects on asynchronous code (`scheduleMicrotask`)

We probably don't have that covered anywhere.

**Comment by jmesserly** [comment]
Good point. I guess most `async` functions would not complete immediately because they're usually doing something :). Also DDC does call `then` asynchronously, so you can't really observe this without zones.

Will send out fix shortly

**Comment by matanlurey** [comment]
Thanks. If you can do a google3 CL for me, I'd be happy to run global test reporting for you!

**Comment by jmesserly** [comment]
Is this something that needs to be cherry picked? Otherwise I'd usually just wait for the next test run once it lands in the SDK repo. I don't think this one is especially risky. (famous last words)

**Comment by matanlurey** [comment]
I don't think it needs to be cherrypicked. I'd like to help make sure it is not risky though :)

**Comment by jmesserly** [comment]
Hmm apparently async_await_test is failing for DDC if I fix this. However, dart2 VM and dart2js also mark it failing. I'll see if I can fix that test, and send the CL to lrn@ to make sure the new behavior is intended.

**Comment by matanlurey** [comment]
No worries. FWIW, I think the DDC behavior is actually more "correct", in that, there is nothing to do asynchronously, and there is nothing returned, therefore there is no need to do extra work. But I realize async/await has a lot of Dart1 baggage still.

**Comment by jmesserly** [comment]
https://dart-review.googlesource.com/c/sdk/+/58980

**Comment by lrhn** [comment]
tl;dr: Don't depend on `scheduleMicrotask` being called. It's not part of the API of any class or method whether it *calls* `scheduleMicrotask`. The interceptors in `Zone` allows you to intercept the *argument* if the function gets called, but nobody promises when, where or whether it will get called. That's simply too much of a straight-jacked when it comes to platform development. The side effects you do in `scheduleMicrotask` should be kept at those that allow communicating information from the call-site to the eventual invocation of the argument function.

To be clear: We make no guarantees how and when future completion propagates, just that they act as if they do so before the current microtask queue is complete. There is no single "right" behavior. Whether a future completion propagates to another future immediately or asynchronously depends on any number of things, including whether another listener throws (if it does, we stop propagation and let the error reach top-level, then continue propagating in a later microtask).

So, if we don't *need* a microtask, we likely won't schedule one.

In this case, I'd prefer to not delay the completion of the `_Future` with a value. We need to delay completion with an error in the first synchronous part of the async function (and preferably not delay the error on later parts of the invocation), but completing a future with a value is safe to do as early as possible, as long as we don't trigger any listeners, and in this case there cannot be any listeners yet.

Examples:
```dart
Future<int> foo1() async {
  return 42;
}
Future<int> foo2() async {
  await new Future.microtask();
  return 37;
}
main() {
  var f1 = foo1();
  var f2 = foo2();
  foo1().then(print);
  foo2().then(print);
}  
```
This  code is likely to print 37 before 42, and *that's OK*.

What happens is that `foo1()` returns an already completed future.
Then `foo2()` schedules a microtask in the `Future.microtask()`.
Then `f1.then(print)` makes `f1` schedules a microtask to deliver the value.
Then `f2.then(print)` registers a listener on `f2`.
Then the future from `Future.microtask()` completes, the `foo2()` invocation continues at the `await`, completes `f2` with 37, which then immediately calls the registered listener (`print`) with the value 37.
Then the microtask from `f1.then(print)` complets and calls `print` with 42.

**Comment by vsmenon** [comment]
At a minimum, I'd like to see DDC and Dart2JS consistent here to avoid forcing developers to debug timing issues on Dart2JS as much as possible.  Ideally, it'd be good if the VM was also consistent.

What's the best way to get there?

**Comment by jmesserly** [comment]
if the DDC behavior seems better (don't schedule a microtask when it's not necessary), could we just change dart2js to match? IDK how difficult that would be.

**Comment by jmesserly** [comment]
Oh, fixing dart2js would be pretty easy, it uses the exact same code as the VM: https://github.com/dart-lang/sdk/blob/master/sdk/lib/_internal/js_runtime/lib/async_patch.dart#L195

@lrhn ... I guess, it looked to me like someone went out of their way to delay completion for VM/dart2js ... I figured there must've been a reason they implemented the timing like that?

**Comment by jmesserly** [comment]
blocked until we know what the intended behavior is. I have the CL to make DDC match dart2js/VM, but it shouldn't be too hard to make dart2js match DDC, if we prefer to go that direction.

**Comment by vsmenon** [comment]
What's the best thing to do for this milestone?  I'd like to converge DDC and Dart2JS at a minimum.  We can:
1. Land Jenny's CL
2. Change dart2js (fyi @sigmundch - sounds like @jmesserly could do this?) and file a bug on the VM for a later milestone
3. Change dart2js and the VM (not sure we have bandwidth for the latter - fyi @a-siva )

**Comment by dgrove** [comment]
Assigning to @vsmenon to figure out what precisely the next steps are.

**Comment by vsmenon** [comment]
@jmesserly - assigning back to you.

Option 1 (preferred).  Keep DDC behavior.  Fix Dart2JS to match it.  Is this reasonable to do right now?

Option 2.  Land @jmesserly 's CL to delay DDC to match Dart2JS VM.

Longer term (i.e., I don't think this need to block Dart 2 Stable), we should file issues to make all the platforms consistent to prevent issues like the original comment.

Objections?

**Comment by lrhn** [comment]
Being consistent on something that is implementation dependent behavior is nice, but not necessary.
Anyone depending on the specifics here are relying on something we do not promise, and do not promise not to change. If anything, having different behavior actually catches such things.

That said, there is no need to delay the completion with the synchronous value, and I'd prefer option 1 if we do decide to change anything.

**Comment by vsmenon** [comment]
Chatted with @sigmundch - resolution is to delayed DDC for now (option 2) - i.e., land @jmesserly 's CL.  Can revisit post-2 stable if we want to change all the backends.

**Comment by vsmenon** [comment]
@jmesserly 's cl is pending:

https://dart-review.googlesource.com/c/sdk/+/58980

**Comment by vsmenon** [comment]
@jmesserly - are you able to land this?

**Comment by jmesserly** [comment]
yeah, let me update the status file (dart 1 mode was failing the new test)
and land this.
